### PR TITLE
Create Modal integration tile

### PR DESCRIPTION
**Pull request by irfansharif**
### What does this PR do?

Adds an integration tile for https://modal.com.

### Motivation

What inspired you to submit this pull request?

### Review checklist

- [x] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [x] Feature or bugfix has tests
- [x] Git history is clean
- [x] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### CODE REVIEW

**Comment by drichards-87** [comment]
Created DOCS-6479 for the Documentation Team review.

**Review by drichards-87 — APPROVED** [review]
Left you some feedback from the Docs Team. As far as the error in the `manifest.json` file, I'll have to defer to the Agent integrations team as I'm not actually sure how to resolve that. My guess is that the logo for the tile needs to be in the assets > logos folder, but I could be wrong about that.

File: modal/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+as well. We're working on surfacing application logs through shortly.
+
+### Events
+
+Modal does not include any events.
+
+## Uninstallation
+
+Once this integration has been uninstalled, any previous authorizations are
+revoked and logs/metrics are
```
[the remark this answers is not in the document]
Comment (irfansharif) [reply modal/README.md]: Oops, done.

**Comment by drichards-87** [comment]
> Thank you for reviewing!
> 
> > I'll have to defer to the Agent integrations team as I'm not actually sure how to resolve that.
> 
> Naive question: are they already tagged on this PR? It's my first time setting up an integration tile thingy, so unsure what the actual process is.

Yes, I believe they're just working through their PR queue right now. Someone should be taking a look at this shortly.

File: modal/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# Modal
+
+## Overview
+
+[Modal][1] lets you run generative AI models, large-scale batch jobs, job
+queues, and much more. The Datadog integration collects logs and metrics from
```
Comment (anawishnoff) [inline modal/README.md:6]:
```suggestion
queues, and much more. This integration collects logs and metrics from
```

File: modal/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+your applications for you to observe directly in your Datadog account.
```
Comment (anawishnoff) [inline modal/README.md:7]: Can you add a bit more context here about what data is  being sent in/what it represents, and why it's valuable for customers?

File: modal/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+## Setup
+
+### Installation
+
+To set up the Modal integration:
+
+1. On the [Datadog Integrations page][2], search for the **Modal** tile and
```
Comment (anawishnoff) [inline modal/README.md:15]:
```suggestion
1. Navigate to the [**Modal** tile][2] on the Datadog Integrations page and
```

File: modal/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+2. Ensure that all API keys associated with this integration have been disabled
+   by searching for the integration name on the [API Keys page][4].
+
+## Troubleshooting
+
+Need help? Contact [Modal support][3].
+
+[1]: https://modal.com
+[2]: https://app.datadoghq.com/integrations
```
Comment (anawishnoff) [inline modal/README.md:56]:
```suggestion
[2]: https://app.datadoghq.com/integrations?integrationId=modal
```

File: modal/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+3. Click the **Authorize** button to complete setup.
+
+Logs and metrics from your Modal apps should now start appearing in Datadog.
+
+## Data Collected
+
+### Metrics
+
+Modal collects utilization metrics for applications/functions that run on Modal.
+You can view these metrics on your [Modal][1] dashboard.
```
Comment (anawishnoff) [inline modal/README.md:31]: It may be more helpful to mention where users can find these metrics and other data within Datadog, rather than within Modal.

File: modal/assets/dashboards/modal_overview.json
Comment (anawishnoff) [inline modal/assets/dashboards/modal_overview.json:1]: Dashboard is looking good so far. Can you also please add a description, and add titles to your widgets that describe what they are showing?

File: modal/manifest.json [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    "overview": "README.md#Overview",
+    "configuration": "README.md#Setup",
+    "support": "README.md#Troubleshooting",
+    "changelog": "CHANGELOG.md",
+    "description": "Collect logs and metrics for your Modal applications",
+    "title": "Modal",
+    "media": [
+      {
+        "media_type": "image",
+        "caption": "Modal logo (svg)",
```
Comment (anawishnoff) [inline modal/manifest.json:16]: For the media carousel, we recommend adding screenshots of your dashboard and your platform to give users an idea of what they'll be receiving. Your logo is automatically displayed on the tile.

**Review by anawishnoff — CHANGES_REQUESTED** [review]
First few comments from the product side.
Can you also please specify which image you'd like to use for your logo? Thanks!

File: modal/assets/dashboards/modal_overview.json
Comment (irfansharif) [reply modal/assets/dashboards/modal_overview.json:1]: Done.

**Review by irfansharif — COMMENTED** [review]
> Can you also please specify which image you'd like to use for your logo? Thanks!

The one under `images/modal-logo-datadog.png`, please.

File: modal/manifest.json [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      "Category::Metrics",
+      "Category::Log Collection",
+      "Submitted Data Type::Metrics",
+      "Submitted Data Type::Logs",
+      "Offering::Integration"
+    ]
+  },
+  "assets": {
+    "dashboards": {
+      "Modal": "assets/dashboards/modal_overview.json"
```
Comment (anawishnoff) [inline modal/manifest.json:34]:
```suggestion
      "Modal": "assets/dashboards/modal.json"
```

File: modal/manifest.json [hunk 9]
```diff
@@ -0,0 +1,59 @@
+{
+  "manifest_version": "2.0.0",
+  "app_uuid": "9567cd54-7fac-46d8-84ce-0b00cef6937e",
+  "app_id": "modal",
+  "display_on_public_website": true,
+  "tile": {
+    "overview": "README.md#Overview",
+    "configuration": "README.md#Setup",
+    "support": "README.md#Troubleshooting",
```
Comment (anawishnoff) [inline modal/manifest.json:9]:
```suggestion
    "support": "README.md#Troubleshooting",
     "uninstallation": "README.md#Uninstallation",
```

File: modal/manifest.json [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      "Modal": "assets/dashboards/modal_overview.json"
+    }, 
+    "integration": {
+      "source_type_name": "Modal",
+      "events": {
+        "creates_events": false
+      },
+      "metrics": {
+        "metadata_path": "metadata.csv",
+        "check": [],
```
Comment (anawishnoff) [inline modal/manifest.json:43]: Please add a check metric. The purpose of the check metrics is that if this metric is received, Datadog will consider the Modal integration to be properly working.

**Review by anawishnoff — CHANGES_REQUESTED** [review]
A few more comments from my end, but otherwise looking good

File: modal/README.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
+# Modal
+
+## Overview
+
+[Modal][1] lets you run generative AI models, large-scale batch jobs, job
+queues, and much more. This integration collects metrics (CPU, memory
+or GPU use for example) and logs (stdout/stderr logging, audit logs) from your
+applications for you to observe directly in your Datadog account.
```
Comment (emarsha94) [inline modal/README.md:8]: The overview should include: 

1. What does your product do (in 1-2 sentences)? This has been done. 
2. What will customers get from this integration, and why is it valuable to them? This should be clear to new customers who are looking at the tile. 
3. What types of data will your integration send in or pull out, **and what's the value of that data?**

File: modal/README.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+## Data Collected
+
+### Metrics
+
+Modal collects GPU, CPU, and memory utilization metrics for
+applications/functions that run on Modal.
+
+### Logs
+
+Modal collects audit logs and all application logs (all stdout/stderr logging).
```
Comment (emarsha94) [inline modal/README.md:36]: To confirm, the integration is sending logs to Datadog?

**Review by emarsha94 — CHANGES_REQUESTED** [review]

File: modal/README.md (hunk 12 above)
Comment (irfansharif) [reply modal/README.md:36]: Yes.

File: modal/README.md (hunk 11 above)
Comment (irfansharif) [reply modal/README.md:8]: Appended a few sentences and re-edited the overview text. It's very general sounding given "modal applications" are effectively any python program, so for instance if a program logged "<N> tasks completed" and then set up a datadog filter for N and graphed that overtime, that's valuable to the author of the application. It's specific to Modal in that we're running the code for you, and emitting those logs to your datadog account, but it's valuable in anyway you choose to use it.

File: modal/manifest.json (hunk 10 above)
Comment (irfansharif) [reply modal/manifest.json:44]: I'm not sure what is appropriate as a check metric for this log-only integration. We're not emitting metrics to datadog yet, simply logs through the log-http endpoint.

Comment (anawishnoff) [reply modal/manifest.json:43]: Ah i see. If this integration is not submitting metrics, you can remove this `metrics` object

File: modal/README.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
+# Modal
+
+## Overview
+
+[Modal][1] lets you run generative AI models, large-scale batch jobs, job
```
Comment (anawishnoff) [inline modal/README.md:5]: Might be helpful to be specific here about how Modal runs the infrastructure to power these jobs

File: modal/README.md [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+queues, and much more. This integration collects metrics (CPU, memory
+or GPU use for example) and logs (stdout/stderr logging from your modal
+applications, or audit logs for your account), for you to observe the data
+directly in your Datadog account. This lets you visualize through Datadog
+dashboards metrics about your modal applications, and also set up alerts.
```
Comment (anawishnoff) [inline modal/README.md:10]:
```suggestion
queues, and much more. This integration collects metrics (CPU, memory
or GPU use for example) and logs (stdout/stderr logging from your modal
applications, or audit logs for your account) that you can visualize through Datadog
dashboards and set up alerts for with Datadog monitors.
```

File: modal/manifest.json (hunk 9 above)
Comment (anawishnoff) [reply modal/manifest.json:9]: This change should be implemented as well

File: modal/manifest.json [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  "tile": {
+    "overview": "README.md#Overview",
+    "configuration": "README.md#Setup",
+    "support": "README.md#Troubleshooting",
+    "changelog": "CHANGELOG.md",
+    "description": "Collect logs and metrics for your Modal applications",
+    "title": "Modal",
+    "media": [
+      {
+        "media_type": "image",
```
Comment (anawishnoff) [inline modal/manifest.json:15]: We recommend also adding an image of your Datadog dashboard

File: modal/manifest.json [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+        "image_url": "images/modal-product-hero.jpg"
+      }
+    ],
+    "classifier_tags": [
+      "Supported OS::Linux",
+      "Supported OS::Windows",
+      "Supported OS::macOS",
+      "Category::Cloud",
+      "Category::Metrics",
+      "Category::Log Collection",
```
Comment (anawishnoff) [inline modal/manifest.json:26]:
```suggestion
      "Category::Log Collection",
      "Category::AI/ML",
```

**Review by anawishnoff — CHANGES_REQUESTED** [review]
A few more suggestions on my end

File: modal/manifest.json (hunk 10 above)
Comment (anawishnoff) [reply modal/manifest.json:43]: @JoshPatel13 can you confirm ?

File: modal/manifest.json [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+      "Category::AI/ML",
+      "Submitted Data Type::Metrics",
+      "Submitted Data Type::Logs",
+      "Offering::Integration"
+    ]
+  },
+  "assets": {
+    "dashboards": {
+      "Modal": "assets/dashboards/modal_overview.json"
```
Comment (anawishnoff) [inline modal/manifest.json:35]: This looks like it's using the wrong path to the dashboard file

```suggestion
      "Modal": "assets/dashboards/modal.json"
```

File: modal/manifest.json (hunk 10 above)
Comment (JoshPatel13) [reply modal/manifest.json:44]: In this case remove the key and value (``` "metrics": {
        "metadata_path": "metadata.csv",
        "check": [],
        "prefix": "modal."
      },```)

Comment (irfansharif) [reply modal/manifest.json:45]: Done.

File: modal/README.md (hunk 13 above)
Comment (irfansharif) [reply modal/README.md:5]: It's feels slightly wordy to be specific here - we make use our custom built containerization tech and filesystems to do so - details we generally don't make apparent to users, hoping it's all seamless. Given we've linked to the modal homepage (where the docs talk through some of these details), what do you think leaving it as is?

**Review by irfansharif — COMMENTED** [review]
Thanks for the review! Took another pass.

File: modal/README.md (hunk 13 above)
Comment (anawishnoff) [reply modal/README.md:5]: Gotcha - definitely dont have to do go into details about the custom containerization, but I think it would add context to the idea that "Modal lets you run xyz" just to explain how it does that. Fine leaving as-is if you think this is descriptive enough!

File: modal/manifest.json (hunk 15 above)
Comment (anawishnoff) [reply modal/manifest.json:16]: For this image, we ask as well that you display dummy data in the dashboard to avoid the No Data state

File: modal/metadata.csv [hunk 18]
```diff
@@ -0,0 +1 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name,curated_metric
```
Comment (anawishnoff) [inline modal/metadata.csv:1]: This file should be updated with all of the metrics that Modal plans to send into DAtadog

File: modal/README.md (hunk 13 above)
Comment (irfansharif) [reply modal/README.md:5]: Ack, I'll leave it as is, but mostly because I'm finding it hard to phrase it too differently without losing folks.

File: modal/metadata.csv (hunk 18 above)
Comment (irfansharif) [reply modal/metadata.csv:1]: Given that this is a log-only integration, for now at least, what should we add here?

File: modal/manifest.json (hunk 15 above)
Comment (irfansharif) [reply modal/manifest.json:16]: Done.

File: modal/manifest.json [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    "title": "Modal",
+    "media": [
+      {
+        "media_type": "image",
+        "caption": "Modal dashboard",
+        "image_url": "images/modal-product-hero.jpg"
+      },
+      {
+        "media_type": "image",
+        "caption": "Modal datadog dashboard",
```
Comment (anawishnoff) [inline modal/manifest.json:22]:
```suggestion
        "caption": "See audit logs, function logs, and utilization metrics in the Modal Datadog dashboard",
```

File: modal/manifest.json [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    "configuration": "README.md#Setup",
+    "support": "README.md#Troubleshooting",
+     "uninstallation": "README.md#Uninstallation",
+    "changelog": "CHANGELOG.md",
+    "description": "Collect logs and metrics for your Modal applications",
+    "title": "Modal",
+    "media": [
+      {
+        "media_type": "image",
+        "caption": "Modal dashboard",
```
Comment (anawishnoff) [inline modal/manifest.json:17]: Can you make this a bit more descriptive as to what the UI is showing in this screenshot?

File: modal/manifest.json [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    ],
+    "classifier_tags": [
+      "Supported OS::Linux",
+      "Supported OS::Windows",
+      "Supported OS::macOS",
+      "Category::Cloud",
+      "Category::Metrics",
+      "Category::Log Collection",
+      "Category::AI/ML",
+      "Submitted Data Type::Metrics",
```
Comment (anawishnoff) [inline modal/manifest.json:34]: If this is only submitting logs at first, please remove this

File: modal/metadata.csv (hunk 18 above)
Comment (anawishnoff) [reply modal/metadata.csv:1]: Ah okay,  in that case this should be left as-is

**Review by anawishnoff — CHANGES_REQUESTED** [review]
A few more comments

**Comment by anawishnoff** [comment]
@irfansharif Can you please merge the master branch into your branch? That will allow the CI to run properly

**Comment by anawishnoff** [comment]
@irfansharif Let us know if you have any questions here, thanks!

File: modal/manifest.json (hunk 19 above)
Comment (irfansharif) [reply modal/manifest.json:22]: Done.

**Comment by irfansharif** [comment]
@anawishnoff: looks like I'm missing some sort of `source_type_id`. Not sure how to fill that in.

**Comment by anawishnoff** [comment]
Hey @irfansharif , this is unfortunately a bit of a manual process right now to add the soruce_type_id, but we are working on improving it. For now, please follow the following steps:

1. upgrade `ddev` to the latest version
2. run `ddev create --type logs "Dummy Integration"`
3. open the file `dummy_integration/manifest.json` that was generated
4. copy the value in the `source_type_id` field, and paste that into your actual `manifest.json` file
5. throw away the dummy_integration files

Also, the `auto_install` field in the manifest.json can be set to `true`. Let me know if you have any questions

**Comment by irfansharif** [comment]
Done!

**Comment by anawishnoff** [comment]
Thanks @irfansharif ! It looks like CI is still catching a few issues - you can [view them here](https://github.com/DataDog/integrations-extras/pull/2135/checks?check_run_id=21328229937). Thanks!

**Review by irfansharif — COMMENTED** [review]
Hey @anawishnoff, I'm not sure what to do about the last remaining failing "Validate repository / run / Validate" test. It's saying something about 
the CI configuration not being in sync?

**Comment by anawishnoff** [comment]
> Hey @anawishnoff, I'm not sure what to do about the last remaining failing "Validate repository / run / Validate" test. It's saying something about the CI configuration not being in sync?

@FlorentClarret would you be able to help with this? 🙏

**Comment by steveny91** [comment]
@irfansharif have you tried running `ddev validate ci --sync`? That should fix the current error.

**Review by irfansharif — COMMENTED** [review]
Ah, great, done.

File: modal/manifest.json [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    "configuration": "README.md#Setup",
+    "support": "README.md#Troubleshooting",
+     "uninstallation": "README.md#Uninstallation",
+    "changelog": "CHANGELOG.md",
+    "description": "Collect logs and metrics for your Modal applications",
+    "title": "Modal",
+    "media": [
+      {
+        "media_type": "image",
+        "caption": "See audit logs, function logs, and utilization metrics in the Modal Datadog dashboard",
```
Comment (anawishnoff) [inline modal/manifest.json:17]: I think this caption should go with the image below, and the product hero image should have a more descriptive caption

**Review by anawishnoff — COMMENTED** [review]
Just one comment left, otherwise looks good, thanks!

**Review by anawishnoff — APPROVED** [review]
Thanks! Approved

**Review by steveny91 — APPROVED** [review]

File: modal/manifest.json [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+      "source_type_id": 7324533,
+      "events": {
+        "creates_events": false
+      },
+      "service_checks": {
+        "metadata_path": "assets/service_checks.json"
+      }
+    },
+    "logs": {}
+  },
```
Comment (JoshPatel13) [inline modal/manifest.json:54]: `integration` here needs a field called configuration. Can look like this `"configuration": {},`. Thanks :)

Comment (irfansharif) [reply modal/manifest.json]: Done!

**Comment by JoshPatel13** [comment]
There needs to be a `metadata.csv` file. Even if there are no metrics, the file needs to be present with the headers defined and empty columns.

**Review by irfansharif — COMMENTED** [review]
Re-added, but BTW, I removed it after CI failures told me to, following the instructions here: https://github.com/DataDog/integrations-extras/pull/2135#discussion_r1443384692.

**Edit:** looks like it's passing now - I'm likely misremembering.

**Comment by JoshPatel13** [comment]
One final thing, could you merge master into your branch and repush.

File: cockroachdb_dedicated/manifest.json [hunk 24]
```diff
@@ -50,4 +50,4 @@
       "source": "cockroach-cloud"
     }
   }
-}
\ No newline at end of file
+}
```
Comment (JoshPatel13) [inline cockroachdb_dedicated/manifest.json:53]: Also, could this change in a different directory be removed.

Comment (JoshPatel13) [inline cockroachdb_dedicated/manifest.json:53]: Could you please undo this change in a different directory? Thanks :)

Comment (irfansharif) [reply cockroachdb_dedicated/manifest.json]: Oops, done. It was my previous employer haha, so that's where I was cargo-culting from.

**Review by JoshPatel13 — APPROVED** [review]

**Review by emarsha94 — APPROVED** [review]

File: modal/CHANGELOG.md [hunk 25]
```diff
@@ -0,0 +1,7 @@
+# CHANGELOG - Modal
+
+## 1.0.0 / 2023-10-17
```
Comment (emarsha94) [inline modal/CHANGELOG.md:3]:
```suggestion
## 1.0.0 / 2024-02-29
```

File: modal/manifest.json [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+  "author": {
+    "support_email": "support@modal.com",
+    "name": "Modal Labs",
+    "homepage": "https://modal.com",
+    "sales_email": "support@modal.com"
+  },
+  "oauth": {}
```
Comment (JoshPatel13) [inline modal/manifest.json:67]:
```suggestion
  }
```

Remove this oauth field as it isn't being used here. Thanks :)
